### PR TITLE
fix: Create GitHub comment as table

### PR DESCRIPTION
**Pull request by novakzaballa**
Thanks for submitting a PR! Please check the boxes below:

- [x] I have run [`pre-commit`](https://docs.flagsmith.com/platform/contributing#pre-commit) to check linting
- [x] I have added information to `docs/` if required so people know about the feature!
- [x] I have filled in the "Changes" section below?
- [x] I have filled in the "How did you test this code" section below?
- [x] I have used a [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) title for this Pull Request

## Changes

- Improve GitHub comment to look like a table

image

## How did you test this code?

- Some unit tests updated.
- Link an external resource manually.

### CODE REVIEW

File: api/integrations/github/constants.py [hunk 1]
```diff
@@ -1,8 +1,14 @@
 GITHUB_API_URL = "https://api.github.com/"
 GITHUB_API_VERSION = "2022-11-28"
 
-LINK_FEATURE_TEXT = "### This pull request is linked to a Flagsmith Feature (`%s`):\n"
+LINK_FEATURE_TITLE = """**Flagsmith feature linked:** `%s`
+Default Values:\n"""
+FEATURE_TABLE_HEADER = """| Environment | Enabled | Value | Type | Updated (UTC) |
```
Comment (khvn26) [inline api/integrations/github/constants.py]:
```suggestion
FEATURE_TABLE_HEADER = """| Environment | Enabled | Value | Type | Last Updated (UTC) |
```

This way, we won't have to prepend "Last Updated" to every timestamp in this column.

File: api/integrations/github/github.py [hunk 2]
```diff
@@ -81,22 +89,37 @@ def generate_body_comment(
     last_updated_string = LAST_UPDATED_FEATURE_TEXT % (
         datetime.datetime.now().strftime("%dth %b %Y %I:%M%p")
     )
```
Comment (khvn26) [inline api/integrations/github/github.py]: I'm not a fan of the custom datetime format.

Let's consider either sticking to ISO8601 or using Django locale, i.e.:

```python
from django.utils.formats import get_format

datetime_format = get_format('DATETIME_INPUT_FORMATS')[0]
```

File: api/integrations/github/github.py [hunk 3]
```diff
@@ -81,22 +89,37 @@ def generate_body_comment(
...
+        if v.get("segment_name") is not None and v["segment_name"] != last_segment_name:
+            result += "\n" + LINK_SEGMENT_TITLE % (v["segment_name"])
+            last_segment_name = v["segment_name"]
+            result += FEATURE_TABLE_HEADER
+        table_row = FEATURE_TABLE_ROW % (
+            v["environment_name"],
+            environment_link_url,
+            "✅ Enabled" if v["feature_value"] else "❌ Disabled",
+            feature_value if feature_value else "",
+            feature_value_type,
```
Comment (khvn26) [inline api/integrations/github/github.py]: I don't think we should surface feature value type here:

1. We plan to move away from feature value types to only support strings in the future.
2. String value type awkwardly gets shown as "Unicode" which, ideally, should be replaced by "String" but it's messy and I'd rather skip the whole ordeal.

File: api/integrations/github/github.py [hunk 4]
```diff
@@ -81,22 +89,37 @@ def generate_body_comment(
...
+        )
+        if v.get("segment_name") is not None and v["segment_name"] != last_segment_name:
+            result += "\n" + LINK_SEGMENT_TITLE % (v["segment_name"])
+            last_segment_name = v["segment_name"]
+            result += FEATURE_TABLE_HEADER
+        table_row = FEATURE_TABLE_ROW % (
+            v["environment_name"],
+            environment_link_url,
+            "✅ Enabled" if v["feature_value"] else "❌ Disabled",
+            feature_value if feature_value else "",
```
Comment (khvn26) [inline api/integrations/github/github.py]: Empty feature value is currently displayed as two backticks (``). I'd like to avoid this.

**Review by khvn26 — CHANGES_REQUESTED** [review]

File: api/integrations/github/github.py [hunk 5]
```diff
@@ -81,22 +89,37 @@ def generate_body_comment(
...
+            tab,
+        )
+        if v.get("segment_name") is not None and v["segment_name"] != last_segment_name:
+            result += "\n" + LINK_SEGMENT_TITLE % (v["segment_name"])
+            last_segment_name = v["segment_name"]
+            result += FEATURE_TABLE_HEADER
+        table_row = FEATURE_TABLE_ROW % (
+            v["environment_name"],
+            environment_link_url,
+            "✅ Enabled" if v["feature_value"] else "❌ Disabled",
```
Comment (khvn26) [inline api/integrations/github/github.py]: I don't think the logic here is correct. Disabled feature states can have values.

I'm not entirely sure where the feature state data comes from, but, assuming it has the "enabled" key, the code should be:

```suggestion
            "✅ Enabled" if v["enabled"] else "❌ Disabled",
```

**Comment by khvn26** [comment]
Please cover the code reported as uncovered with tests, or add a `# pragma: no cover` annotation, preferably with an explanation why.

File: api/integrations/github/constants.py (hunk 1 above)
Comment (novakzaballa) [reply api/integrations/github/constants.py]: Done.

File: api/integrations/github/github.py (hunk 2 above)
Comment (novakzaballa) [reply api/integrations/github/github.py]: Done.

File: api/integrations/github/github.py (hunk 4 above)
Comment (novakzaballa) [reply api/integrations/github/github.py]: Done.

File: api/integrations/github/github.py (hunk 3 above)
Comment (novakzaballa) [reply api/integrations/github/github.py]: Done.

File: api/integrations/github/github.py (hunk 5 above)
Comment (novakzaballa) [reply api/integrations/github/github.py]: The logic is correct, but the variable name is confusing. Variable renamed

**Comment by novakzaballa** [comment]
> Please cover the code reported as uncovered with tests, or add a `# pragma: no cover` annotation, preferably with an explanation why.

I ended up adding the coverage. Let's keep this coverage while it is possible and is worth it.

File: api/tests/unit/features/test_unit_feature_external_resources_views.py [hunk 6]
```diff
@@ -41,20 +40,31 @@ def json(self):
...
+
+    expected_comment_body = (
+        "**Flagsmith feature linked:** `feature_with_value`\n"
+        + "Default Values:\n"
+        + "| Environment | Enabled | Value | Last Updated (UTC) |\n"
+        + "| :--- | :----- | :------ | :------ |\n"
+        + f"| [Test Environment](https://example.com/project/{project.id}/"
+        + f"environment/{environment.api_key}/features?feature={feature_with_value.id}&tab=value) "
+        + "| ❌ Disabled | `value` | 2024-01-01 00:00:00 |\n"
+    )
```
Comment (khvn26) [inline api/tests/unit/features/test_unit_feature_external_resources_views.py]: This is nice! nit: Python supports multiline strings so you could do this:

```suggestion
    expected_comment_body = (
        "**Flagsmith feature linked:** `feature_with_value`\n"
        "Default Values:\n"
        "| Environment | Enabled | Value | Last Updated (UTC) |\n"
        "| :--- | :----- | :------ | :------ |\n"
        f"| [Test Environment](https://example.com/project/{project.id}/"
        f"environment/{environment.api_key}/features?feature={feature_with_value.id}&tab=value) "
        "| ❌ Disabled | `value` | 2024-01-01 00:00:00 |\n"
    )
```

Looks a bit nicer this way, IMO.

File: api/integrations/github/github.py [hunk 7]
```diff
@@ -78,25 +86,38 @@ def generate_body_comment(
...
+            get_current_site_url(),
+            project_id,
+            v.get("environment_api_key"),
+            feature_id,
+            tab,
+        )
+        if v.get("segment_name") is not None and v["segment_name"] != last_segment_name:
+            result += "\n" + LINK_SEGMENT_TITLE % (v["segment_name"])
+            last_segment_name = v["segment_name"]
+            result += FEATURE_TABLE_HEADER
```
Comment (khvn26) [inline api/integrations/github/github.py:105]: We should add a segment to the test or slap a `# pragma: no cover` on L108.

File: api/tests/unit/features/test_unit_feature_external_resources_views.py (hunk 6 above)
Comment (novakzaballa) [reply api/tests/unit/features/test_unit_feature_external_resources_views.py]: Done.

File: api/integrations/github/github.py (hunk 7 above)
Comment (novakzaballa) [reply api/integrations/github/github.py:105]: Done.

**Review by khvn26 — APPROVED** [review]

**Pull request merged**